SNI extension

ashfall opened the pull request:
Implementing as specified in https://tools.ietf.org/html/rfc6066#section-3

ashfall commented on tls/test/test_hello_message.py:
```diff
@@ -173,6 +180,19 @@ def test_as_bytes_client_hello_cipher_suites(self):
+        assert len(record.extensions[0].data) == 1
+        server_name_list = record.extensions[0].data
+        assert server_name_list[0].name_type == enums.NameType.HOST_NAME
+        assert server_name_list[0].name.data == 'localhost'
```
I'm not very happy about doing `server_name_list[0].name.data` and would rather `localhost` be `server_name_list[0].name`, i.e. not have the HostName container as a thing. From the RFC (https://tools.ietf.org/html/rfc6066):

```
      struct {
          NameType name_type;
          select (name_type) {
              case host_name: HostName;
          } name;
      } ServerName;

      enum {
          host_name(0), (255)
      } NameType;

      opaque HostName<1..2^16-1>;

      struct {
          ServerName server_name_list<1..2^16-1>
      } ServerNameList;
```

ashfall commented:
I'm not very happy about doing `server_name_list[0].name.data` and would rather `'localhost'` be `server_name_list[0].name`, i.e. not have the `HostName` container as a thing. From the RFC (https://tools.ietf.org/html/rfc6066):

```
      struct {
          NameType name_type;
          select (name_type) {
              case host_name: HostName;
          } name;
      } ServerName;

      enum {
          host_name(0), (255)
      } NameType;

      opaque HostName<1..2^16-1>;

      struct {
          ServerName server_name_list<1..2^16-1>
      } ServerNameList;
```

Thoughts?

markrwilliams commented on tls/_constructs.py:
```diff
@@ -67,6 +68,25 @@
+        {
+            enums.NameType.HOST_NAME: HostName
+        }
+    )
```
[Agreed!](#106)

markrwilliams commented on tls/_constructs.py:
```diff
@@ -67,6 +68,25 @@
     Array(lambda ctx: ctx.length, UBInt8("compression_methods"))
 )
 
+HostName = Struct(
```
From @ashfall's [comment on this PR](https://github.com/pyca/tls/pull/105#issuecomment-261179286):

> I'm not very happy about doing server_name_list[0].name.data and would rather 'localhost' be server_name_list[0].name, i.e. not have the HostName container as a thing.

I think you can remove the `Struct` and name the `PrefixedBytes` `hostname`:

```python
HostName = PrefixedBytes("hostname", UBInt16("length"))
```

But I haven't tested it.

markrwilliams commented on tls/test/test_hello_message.py:
```diff
@@ -236,14 +256,11 @@ def test_parse_server_hello(self):
-        :func:`parse_server_hello` returns an instance of :class:`ServerHello`
-        with extensions, when the extension bytes are present in the input.
+        :func:`parse_server_hello` fails to parse when
+        SIGNATURE_ALGORITHMS extention bytes are present in the packet
```
Looks like the docs build failure is due to `extention`:

https://travis-ci.org/pyca/tls/jobs/176794154#L433

markrwilliams commented on tls/hello_message.py:
```diff
@@ -72,6 +91,9 @@ def from_bytes(cls, bytes):
         :return: ClientHello object.
         """
         construct = _constructs.ClientHello.parse(bytes)
+        if any(extension.type not in cls.allowed_extensions
```
This was an excellent catch!  Should `as_bytes` also have this check, so that it's not possible to serialize a hello message that contains invalid extensions?

ashfall commented:
@markrwilliams Thanks for the review, updates coming!

I have a question for you about the `EnumSwitch` API. In order to[ satisfy the coverage requirement](https://coveralls.io/builds/8879968/source?filename=tls%2Fhello_message.py#L96), I was writing a test with an unsupported extension in `ClientHello`:

```
    def test_client_hello_fails_with_unsupported_extension(self):
        """
        :py:func:`tls.hello_message.ClientHello` does not parse a packet
        with an unsupported extension, and raises an error.
        """
        server_certificate_type_extension_data = (
            b'\x00\x14'  # Extension Type: Server Certificate Type
            b'\x00\x00'  # Length
            b''  # Data
        )

        client_hello_packet = self.common_client_hello_data + (
            b'\x00\x04'  # extensions length
        ) + server_certificate_type_extension_data

        with pytest.raises(UnsupportedExtensionException):
            ClientHello.from_bytes(
                client_hello_packet
            )
```

Since we haven't implemented the `server_certificate_type` extension yet, `Extension` raised a `SwitchError: no default case defined` on parsing `server_certificate_type_extension_data`. 

To define a default, `EnumSwitch` [says](https://github.com/ashfall/tls/blob/master/tls/_common/_constructs.py#L152): 
```    :param value_choices: A dictionary that maps members of
        `type_enum` to subconstructs.  This follows
        :py:func:`construct.core.Switch`'s API, so ``_default_`` will
        match any members without an explicit mapping.
    :type value_choices: :py:class:`dict`
```
and it was unclear to me how/where to provide the ``__default__``. 

To make my test pass, I hacked around a bit and made the following changes:

I modified `EnumSwitch` slightly, a rough implementation is below (the if-else is ugly, I should clean that up):
```
def EnumSwitch(type_field, type_enum, value_field, value_choices, default=None): 
    if not default:
        return (EnumClass(type_field, type_enum),
                construct.Switch(value_field,
                                 operator.attrgetter(type_field.name),
                                 value_choices))
    else:
        return (EnumClass(type_field, type_enum),
                construct.Switch(value_field,
                                 operator.attrgetter(type_field.name),
                                 value_choices,
                                 default=default))
```

And added a default `construct.Pass` to the `Extension` construct:
```
Extension = Struct(
    "extensions",
    *EnumSwitch(
        type_field=UBInt16("type"),
        type_enum=enums.ExtensionType,
        value_field="data",
        value_choices={
            enums.ExtensionType.SERVER_NAME: Opaque(ServerNameList),
            enums.ExtensionType.SIGNATURE_ALGORITHMS: Opaque(
                SupportedSignatureAlgorithms
            ),
        },
        default=Pass,
    )
)
```

This made `test_client_hello_fails_with_unsupported_extension` pass.

All the hackery aside, is there an existing way to provide default values to `EnumSwitch`?

markrwilliams commented:
@ashfall This looks great!  I'm going to merge.  We can address any lingering concerns in subsequent PRs.

Thanks so much for implementing this!
